### DISCUSSION TITLE
Filtred data received from MessageBus in SubcribeTopic by resourceName

### DISCUSSION THREAD

**Discussion by MhamedHr** [edgexfoundry/community#136] [Q&A]
I am attempting to filter data received from the MessageBus in the SubcribeTopic based on the **resourceName**. However, I encountered an issue where the documentation does not provide a method to achieve this directly. Currently, we can only filter by sourceName using the format : **`edgex/events/device/<profile-name>/<device-name>/<source-name>`**.

Is it possible to retrieve just a specific **resourceName** within a device profile?"

**Comment by cloudxxx8** [comment]
You can use `edgex/events/device/+/+/<resource-name>`
`+` is the wildcard

**Reply by MhamedHr** [reply]
For me, I work with Redis using '#' instead of '+'. Moreover, I tried to use `**edgex/events/device/#/#/<resourceName>**`, but it doesn't work either."
I think it's not possible to filter by **resourceName** on **SubdribeTopics**

**Reply by cloudxxx8** [reply]
No, EdgeX message bus is an abstraction of the real message bus implementation. We use a common way to define the topic, and convert to different implementation.  To use pure Redis pub/sub, you should not use `/` in the channel definition.
Please just try my answer.

**Reply by MhamedHr** [reply]
I have already tried, but if possible, could you give me an example of how to read **resourceName** from Topics using Redis?

**Reply by cloudxxx8** [reply]
oh, I guess you are developing your own application to subscribe the messages.
my answer can only work for App Service or any service using go-mod-messaging.
To use v3.0 there is one more layer
`edgex/events/device/<device-service-name>/<profile-name>/<device-name>/<source-name>`
It should be `edgex/events/device/+/+/+/<resource-name>`

For the pure Redis, please read the Redis document.  It's out of the scope of this community.

**Reply by MhamedHr** [reply]
Thank you for your response. I have another question if possible, how can we secure the messageBus to prevent application services from reading all resourceNames in the same profile device? In other words, how can we restrict app services to only read a specific resourceName?"

**Reply by cloudxxx8** [reply]
It's just like my original answer.  Set the subscribe topic of trigger to `edgex/events/device/+/+/+/<resource-name>`
```
Trigger:
  SubscribeTopics: "edgex/events/device/+/+/+/<resource-name>"
```
Like this example to customize the SubscribeTopics
https://github.com/edgexfoundry/app-service-configurable/blob/162d3a06d6a77858f9267e44f258c8258eccc8be/res/metrics-influxdb/configuration.yaml#L49

**Comment by lenny-goodell** [comment]
Note that `source-name` is either the resource or command name that triggers the Event to be created.
https://docs.edgexfoundry.org/3.0/microservices/application/Triggers/#filter-by-topics
So as @cloudxxx8 pointed out above for your use case use the  resource name for `source-name`
